### PR TITLE
crystfel: 0.10.2 -> 0.11.0

### PR DESCRIPTION
**Pull request by pmiddend**
## Description of changes

CrystFEL 0.11.0 was released recently. I'm working at the same institution and took up maintenance of the NixOS package. Things of note:

- There is a new dependency, millepede-ii (see [here](https://gitlab.desy.de/claus.kleinwort/millepede-ii/-/tree/V04-13-06?ref_type=tags), also developed at DESY, LGPL-licensed), which is used in CrystFEL in a simple system call, so I added the path to this dependency to `PATH` for all CrystFEL binaries.
- One of the patches (OSX-related) had to be updated, it didn't apply anymore, other patches have been mainlined.
- This includes a small bugfix (see the new patch).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by pmiddend** [comment]
@ofborg build crystfel

**Review by rhendric — APPROVED** [review]

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 1]
```diff
@@ -121,10 +122,10 @@ let
 
   pinkIndexer = stdenv.mkDerivation rec {
     pname = "pinkindexer";
-    version = "5d4e016941eb2a9e50a10df96ded7ff1e2464503";
+    version = "15caa21191e27e989b750b29566e4379bc5cd21a";
     src = fetchurl {
       url = "https://gitlab.desy.de/thomas.white/${pname}/-/archive/${version}/${pname}-${version}.tar.gz";
```
Comment (AndersonTorres) [inline pkgs/applications/science/physics/crystfel/default.nix:127]: inline `pname`

#277994

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 2]
```diff
@@ -202,11 +220,7 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./link-to-argp-standalone-if-needed.patch
```
Comment (AndersonTorres) [inline pkgs/applications/science/physics/crystfel/default.nix:222]: Add a comment explaining the patches.

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 3]
```diff
@@ -216,9 +230,12 @@ stdenv.mkDerivation rec {
     sed -i -e 's#execlp("mosflm"#execl("${mosflm}/bin/mosflm"#' libcrystfel/src/indexers/mosflm.c;
   '';
 
+  mesonBuildType = "debugoptimized";
```
Comment (AndersonTorres) [inline pkgs/applications/science/physics/crystfel/default.nix:233]: Why?

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 4]
```diff
@@ -216,9 +230,12 @@ stdenv.mkDerivation rec {
...
+  dontStrip = true;
+
   postInstall = lib.optionalString withBitshuffle ''
     for file in $out/bin/*; do
-      wrapProgram $file --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins
+      wrapProgram $file --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins --prefix PATH ":" ${lib.makeBinPath [ millepede-ii ]}
```
Comment (AndersonTorres) [inline pkgs/applications/science/physics/crystfel/default.nix:238]:
```suggestion
      wrapProgram $file \
        --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins \
        --prefix PATH ":" ${lib.makeBinPath [ millepede-ii ]}
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 1 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix:127]: Didn't know about this! Fixed in all places, not just here.

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 2 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix:224]: Good point, added some comments.

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 3 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix:237]: Debugging reasons. I removed it.

**Comment by pmiddend** [comment]
I'm finished with this round of changes.

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 5]
```diff
@@ -169,13 +170,29 @@ let
       "-DENABLE_BZIP2_PLUGIN=yes"
     ];
   };
+
+  millepede-ii = stdenv.mkDerivation rec {
+    pname = "millepede-ii";
+    version = "V04-13-06";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:176]:
```suggestion
    version = "04-13-06";
```
we strip version prefixes like v, REL or the package name.

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 6]
```diff
@@ -169,13 +170,29 @@ let
...
+    src = fetchurl {
+      url = "https://gitlab.desy.de/claus.kleinwort/millepede-ii/-/archive/${version}/millepede-ii-${version}.tar.gz";
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:178]:
```suggestion
      url = "https://gitlab.desy.de/claus.kleinwort/millepede-ii/-/archive/V${version}/millepede-ii-V${version}.tar.gz";
```

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 7]
```diff
@@ -169,13 +170,29 @@ let
...
+      hash = "sha256-aFoo8AGBsUEN2u3AmnSpTqJ6JeNV6j9vkAFTZ34I+sI=";
+    };
+
+    nativeBuildInputs = [ gfortran ];
+    buildInputs = [ zlib ];
+
+    preBuild = ''
+      makeFlagsArray+=("PREFIX=$out")
+    '';
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:187]: Why not use makeFlags directly?

```suggestion
    makeFlags = [
      "PREFIX=$(out)"
    ];
```

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 8]
```diff
@@ -216,9 +234,13 @@ stdenv.mkDerivation rec {
     sed -i -e 's#execlp("mosflm"#execl("${mosflm}/bin/mosflm"#' libcrystfel/src/indexers/mosflm.c;
   '';
 
+  mesonBuildType = "debugoptimized";
+  dontStrip = true;
+
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:239]:
```suggestion
```

File: pkgs/applications/science/physics/crystfel/default.nix [hunk 9]
```diff
@@ -216,9 +234,13 @@ stdenv.mkDerivation rec {
...
   postInstall = lib.optionalString withBitshuffle ''
     for file in $out/bin/*; do
-      wrapProgram $file --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins
+      wrapProgram $file --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins \
+                        --prefix PATH ":" ${lib.makeBinPath [ millepede-ii ]}
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/crystfel/default.nix:243]:
```suggestion
      wrapProgram $file \
        --set HDF5_PLUGIN_PATH ${hdf5-external-filter-plugins}/lib/plugins \
        --prefix PATH ":" ${lib.makeBinPath [ millepede-ii ]}
```

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 5 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Ah, wasn't aware of that. Fixed.

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 8 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Forgot about that. Fixed.

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 7 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Not sure, but seems to work the same, fixed.

File: pkgs/applications/science/physics/crystfel/default.nix (hunk 9 above)
Comment (pmiddend) [reply pkgs/applications/science/physics/crystfel/default.nix]: Okay, reformatted that.

**Comment by pmiddend** [comment]
Second batch of changes finished.

**Pull request merged**